### ISSUE TITLE
Get all Services (name, displayname, status, start type) to write to Influx. 

### ISSUE THREAD

**Issue by Varjotiger** [markwragg/PowerShell-Influx#26]
Hi all,

I'm trying to get all Windows server services to write to InfluxDB but am experiencing a slight issue. I initially tried to do the following (I left the basic metrics in to see if it was writing) until I realised that `$Metrics` does not like nested hash tables.

```powershell
#Influx Database HTTP Address.
$influxDB = "http://localhost:8086"
    # Get all of your metrics.
    While($true){
        $ValuesList = Get-Counter -Counter "\Memory\Available Bytes","\Processor(*)\% Processor Time"
        $MemoryActive = [math]::Round($ValuesList.CounterSamples[0].CookedValue,3) /1024 /1024
        $MemoryFreeMB = [math]::Round($MemoryActive)
        $ProcessorActive = [math]::Round($ValuesList.CounterSamples[1].CookedValue,2)
        $GetWindowsServices = Get-Service | select -Property Name,DisplayName,Status,StartType
        $Metrics = @{
            "CPU Active" = $ProcessorActive
            "Free RAM MB" = $MemoryFreeMB
            "Services" = $GetWindowsServices
    }
    # Send all Metrics to InfluxDB.
    Write-Influx -Measure Server -Tags @{Prod=$env:COMPUTERNAME} -Metrics $Metrics -Database telegraf -Server http://localhost:8086 -Verbose}
```

I was wondering if I could do a `foreach` on the `$GetWindowsServices` hash table to get this to work. I am learning PowerShell fairly quickly but this has me stumped.

All in all I'm hoping that if this works for the Services then I can use the same format for Windows Event Logs.

Thanks

P.s. I'm very new to github and couldn't find how to add this as a request rather than an issue.

**Comment by markwragg** [comment]
I believe Influx supports writing metric values as one of 4 types: `string`, `float`, `integer` or `boolean`, so as you've suggested, you really need to split up the object collection you've got in `$GetWindowsServices` when you write it to Influx. 

It's worth noting also that while `$GetWindowsServices` might seem like a hashtable, its actually a PowerShell object collection.

Here's what I think you want to do:

```powershell
# Influx Database HTTP Address
$influxDB = "http://localhost:8086"

# Get all of your services
$GetWindowsServices = Get-Service

# Iterate over each service and write to Influx individually
ForEach ($Service in $GetWindowsServices) {
    $Metrics = @{
        Name        = $Service.Name
        DisplayName = $Service.DisplayName
        Status      = $Service.Status
        StartType   = $Service.StartType
    }

    Write-Influx -Measure Server -Tags @{Prod=$env:COMPUTERNAME} -Metrics $Metrics -Database telegraf -Server $influxDB -Verbose
}
```

You might want to also consider writing the service name as a tag as well as a metric as that can make it easier to query/filter the data on that value if you wanted to.

Regarding raising this as an issue, that's fine, GitHub doesn't have a concept of requests, everything is just raised as issues and we use labels to differentiate/organise.

**Comment by Varjotiger** [comment]
Hi Mark,

First off, apologies for mentioning hash tables, I had actually started looking into nested hash tables to try and get the information across to InfluxDB.

Thank you VERY much for your response, you have been a massive help.  I now have all Services writing to a table and can play around with appending the entries rather than capturing and writing the Service over and over (Spooler, Spooler, Spooler).  As you advised, I also added tags to the different metrics.

Cheers again mate, I was going round the twist trying to get that to work.

I hope you have a great day!

Mark
